policy/compiler/compiler.go

```diff
@@ -42,38 +43,165 @@ type Compiler struct {
 	reg Registry
 }
 
+// CompileInstance type-checks and validates a parsed representation of a policy instance whose
+// format and validation logic is also determined by policy template referenced in the policy
+// instance 'kind' field.
+func (c *Compiler) CompileInstance(src *model.Source,
+	inst *model.ParsedValue) (*model.Instance, *common.Errors) {
+	return c.newInstanceCompiler(src, inst).compile()
+}
+
 // CompileTemplate type-checks and validates a parsed representation of a policy template.
 func (c *Compiler) CompileTemplate(src *model.Source,
 	tmpl *model.ParsedValue) (*model.Template, *common.Errors) {
-	tc := &templateCompiler{
+	return c.newTemplateCompiler(src, tmpl).compile()
+}
+
+func (c *Compiler) newInstanceCompiler(src *model.Source,
+	inst *model.ParsedValue) *instanceCompiler {
+	dc := c.newDynCompiler(src, inst)
+	dyn := model.NewDynValue(inst.ID, inst.Value)
+	tmplName := dc.mapFieldStringValueOrEmpty(dyn, "kind")
+	tmpl, found := c.reg.FindTemplate(tmplName)
+	if !found {
+		// report an error and return
```

Comment:
Does the error come in a followup PR?

Reply:
Nope, that should come now. :) I'll fix before merge.